engine/buildkit/resources/cpustat.go

```diff
@@ -0,0 +1,145 @@
+package resources
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"dagger.io/dagger/telemetry"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+const (
+	cpuStatFile     = "cpu.stat"
+	cpuPressureFile = "cpu.pressure"
+
+	cpuUsageKey  = "usage_usec"
+	cpuUserKey   = "user_usec"
+	cpuSystemKey = "system_usec"
+)
+
+type cpuStatSampler struct {
+	cpuStatFilePath string
+	commonAttrs     attribute.Set
+
+	cpuUsage  metric.Int64Gauge
+	cpuUser   metric.Int64Gauge
+	cpuSystem metric.Int64Gauge
+}
+
+func newCPUStatSampler(cgroupPath string, meter metric.Meter, commonAttrs attribute.Set) (*cpuStatSampler, error) {
+	s := &cpuStatSampler{
+		cpuStatFilePath: filepath.Join(cgroupPath, cpuStatFile),
+		commonAttrs:     commonAttrs,
+	}
+	var err error
+
+	s.cpuUsage, err = meter.Int64Gauge(telemetry.CPUStatUsage, metric.WithUnit(telemetry.MicrosecondUnitName))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create cpuUsage metric: %w", err)
+	}
+
+	s.cpuUser, err = meter.Int64Gauge(telemetry.CPUStatUser, metric.WithUnit(telemetry.MicrosecondUnitName))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create cpuUser metric: %w", err)
+	}
+
+	s.cpuSystem, err = meter.Int64Gauge(telemetry.CPUStatSystem, metric.WithUnit(telemetry.MicrosecondUnitName))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create cpuSystem metric: %w", err)
+	}
+
+	return s, nil
+}
+
+type cpuStatSample struct {
+	cpuUsage  int64GaugeSample
+	cpuUser   int64GaugeSample
+	cpuSystem int64GaugeSample
+}
+
+func (s *cpuStatSampler) sample(ctx context.Context) error {
+	sample := cpuStatSample{
+		cpuUsage:  newInt64GaugeSample(s.cpuUsage, s.commonAttrs),
+		cpuUser:   newInt64GaugeSample(s.cpuUser, s.commonAttrs),
+		cpuSystem: newInt64GaugeSample(s.cpuSystem, s.commonAttrs),
+	}
+
+	bs, err := os.ReadFile(filepath.Join(s.cpuStatFilePath, cpuStatFile))
+	if err != nil {
+		return fmt.Errorf("failed to read cpu.stat file: %w", err)
+	}
+
+	for key, value := range flatKeyValuesInt64(bs) {
+		switch key {
+		case cpuUsageKey:
+			sample.cpuUsage.add(value)
+		case cpuUserKey:
+			sample.cpuUser.add(value)
+		case cpuSystemKey:
+			sample.cpuSystem.add(value)
+		}
+	}
+
+	sample.cpuUsage.record(ctx)
+	sample.cpuUser.record(ctx)
+	sample.cpuSystem.record(ctx)
+
+	return nil
+}
+
+type cpuPressureSampler struct {
+	cpuPressureFilePath string
+	commonAttrs         attribute.Set
+
+	someTotal metric.Int64Gauge
+	fullTotal metric.Int64Gauge
+}
+
+func newCPUPressureSampler(cgroupPath string, meter metric.Meter, commonAttrs attribute.Set) (*cpuPressureSampler, error) {
+	s := &cpuPressureSampler{
+		cpuPressureFilePath: filepath.Join(cgroupPath, cpuPressureFile),
+		commonAttrs:         commonAttrs,
+	}
+	var err error
+
+	s.someTotal, err = meter.Int64Gauge(telemetry.CPUStatPressureSomeTotal, metric.WithUnit(telemetry.MicrosecondUnitName))
```

Comment:
Passing thought, not blocking: could be worth configuring [`WithDescription`](https://pkg.go.dev/go.opentelemetry.io/otel/metric#WithDescription) for all of these. I'm guessing that will make it all the way to the front-end, where it could be e.g. a tooltip in the web UI.

Reply:
Agreed, added those